c2rust/src/main.rs

```diff
@@ -1,56 +1,154 @@
-use clap::{crate_authors, load_yaml, App, AppSettings, SubCommand};
+use anyhow::anyhow;
+use camino::Utf8Path;
+use clap::{crate_authors, App, AppSettings, Arg};
 use git_testament::{git_testament, render_testament};
+use is_executable::IsExecutable;
+use std::borrow::Cow;
+use std::collections::HashMap;
 use std::env;
 use std::ffi::OsStr;
-use std::process::{exit, Command};
+use std::path::{Path, PathBuf};
+use std::process;
+use std::process::Command;
+use std::str;
 
 git_testament!(TESTAMENT);
 
-fn main() {
-    let subcommand_yamls = [load_yaml!("transpile.yaml")];
-    let matches = App::new("C2Rust")
-        .version(&*render_testament!(TESTAMENT))
-        .author(crate_authors!(", "))
-        .setting(AppSettings::SubcommandRequiredElseHelp)
-        .subcommands(
-            subcommand_yamls
-                .iter()
-                .map(|yaml| SubCommand::from_yaml(yaml)),
-        )
-        .get_matches();
+/// A `c2rust` sub-command.
+struct SubCommand {
+    /// The path to the [`SubCommand`]'s executable,
+    /// if it was found (see [`Self::find_all`]).
+    /// Otherwise [`None`] if it is a known [`SubCommand`] (see [`Self::known`]).
+    path: Option<PathBuf>,
+    /// The name of the [`SubCommand`], i.e. in `c2rust-{name}`.
+    name: Cow<'static, str>,
+}
 
-    let mut os_args = env::args_os();
-    os_args.next(); // Skip executable name
-    let arg_name = os_args.next().and_then(|name| name.into_string().ok());
-    match (&arg_name, matches.subcommand_name()) {
-        (Some(arg_name), Some(subcommand)) if arg_name == subcommand => {
-            invoke_subcommand(subcommand, os_args);
-        }
-        _ => {
-            eprintln!("{:?}", arg_name);
-            panic!("Could not match subcommand");
+impl SubCommand {
+    /// Find all [`SubCommand`]s adjacent to the current (`c2rust`) executable.
+    /// They are of the form `c2rust-{name}`.
+    pub fn find_all() -> anyhow::Result<Vec<Self>> {
+        let c2rust = env::current_exe()?;
+        let c2rust_name: &Utf8Path = c2rust
```

Comment:
What purpose is the Utf8Path serving here?

Reply:
I think I used `camino` and `Utf8Path` instead of just `Path` because `Path` and `OsStr` lack a lot of methods, like `strip_prefix` and `starts_with`, and it's kind of impossible to have that functionality in a cross-platform way.  These names are very unlikely not to be UTF-8, so it's just easier to enforce that with `camino::Utf8Path`.

Reply:
Sure, but if the file is a valid Utf8Path then the it's a valid String, no? Just seems like overkill. It's not necessarily wrong, just fancier than I would expect we would need.

Reply:
My reasoning here is that we shouldn't add dependencies for simple things unless they are _extremely_ convenient or we use it a lot. Managing dependencies, keeping them up-to-date, making sure they are maintained and get security fixes, even just downloading and building an extra dep has a cost. Not necessarily a huge deal for this project, but something to weigh against.

Seems we already use camino though? So we can just keep using it here I guess, even though I'd prefer not to.

Reply:
I agree with the above--I'm not a huge fan of more dependencies but it shouldn't block this PR especially as we are already using `camino` in other c2rust crates. Someone could probably remove all usages of it in without too much work in an independent PR if they wanted.

Reply:
Okay, I can see about removing it, here and in other places.  I think I had initially added it since I was using it much more thoroughly, but this one use here can definitely be removed.

Reply:
I'll remove it in another PR, along with the other locations.

Reply:
@fw-immunant, @rinon, this (removing `camino`) is fixed in #644.